File: plugins/harness/skills/harness-install/assets/maven/harness-maven-plugin/src/main/java/com/ririnto/sinon/harness/rules/fs/FilePresenceRule.java
package com.ririnto.sinon.harness.rules.fs;

import com.ririnto.sinon.harness.rules.HarnessCheckHelper;
import com.ririnto.sinon.harness.rules.HarnessCheckRule;
import com.ririnto.sinon.harness.core.RuleContext;
import com.ririnto.sinon.harness.Finding;

import tools.jackson.databind.JsonNode;
import org.apache.maven.plugin.MojoExecutionException;
import java.nio.file.Path;
import java.util.Collection;

/**
 * Rule that requires specified files to exist in the project.
 */
public enum FilePresenceRule implements HarnessCheckRule {
    INSTANCE;

    private static final String CATEGORY = "filePresence";

    @Override
    public String category() {
        return "filePresence";
    }

    @Override
    public boolean applies(RuleContext ctx) {
        return ctx.manifest().isEnabled(CATEGORY);
    }

    @Override
    public Collection<Finding> validate(RuleContext ctx) throws MojoExecutionException {
        final Path root = ctx.root();
        final JsonNode manifest = ctx.manifest().raw();
        final String severity = HarnessCheckHelper.getSeverity(manifest, CATEGORY);
        return HarnessCheckHelper.extractPaths(manifest.get(CATEGORY).get("parameters").get("paths")).stream()
                .filter(path -> !HarnessCheckHelper.isSafeRegularFile(root, root.resolve(path)))
                .map(path -> Finding.of(severity, CATEGORY, "missing file: " + path))
                .toList();
    }
}

File: plugins/harness/skills/harness-install/assets/gradle/buildSrc/src/main/kotlin/com/ririnto/sinon/harness/rules/fs/FilePresenceRule.kt
package com.ririnto.sinon.harness.rules.fs

import com.ririnto.sinon.harness.ast.HarnessAstResults.Finding
import com.ririnto.sinon.harness.core.RuleContext
import com.ririnto.sinon.harness.core.Severity
import com.ririnto.sinon.harness.rules.HarnessCheckRule
import kotlinx.serialization.json.JsonObject
import java.nio.file.Path
import kotlin.io.path.div
import kotlin.io.path.isRegularFile
import kotlin.io.path.isSymbolicLink

/**
 * Rule that requires specified files to exist.
 *
 * @category
 *   filePresence
 */
object FilePresenceRule : HarnessCheckRule() {
    /**
     * Category identifier for this rule.
     */
    override val category: String = "filePresence"

    override fun validate(ctx: RuleContext): Collection<Finding> {
        (ctx.manifest.categoryObject(category) ?: return emptyList()).get("parameters") as? JsonObject ?: return emptyList()
        return buildList {
            ctx.manifest.stringArray(category, "paths").forEach { path ->
                val p = ctx.root / path
                when {
                    p.isSymbolicLink() && !ctx.isAllowedRootContractSymlink(p) -> {
                        add(Finding(Severity.ERROR, category, "symlink file is not allowed: $path"))
                    }

                    !p.isRegularFile() -> {
                        add(Finding(ctx.manifest.severityOf(category), category, "missing file: $path"))
                    }
                }
            }
        }
    }
}
